torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -232,7 +297,25 @@ void Reducer::mark_variable_ready_dense(VariableIndex index) {
       TORCH_INTERNAL_ASSERT(!grad.is_alias_of(bucket_view));
       TORCH_INTERNAL_ASSERT(grad.device() == bucket_view.device());
       TORCH_INTERNAL_ASSERT(grad.numel() == bucket_view.numel());
-      bucket_view.copy_(grad.view({-1}), /* non_blocking */ true);
+      // AccumulateGrad doesn't HAVE to obey the grad layout contract.
+      // The penalty for disobedience is reduced performance, not numerical death.
+      // Warnings here help diagnose poor DDP performance.
+      if (grad.strides() != bucket_view.strides()) {
+        TORCH_WARN_ONCE("Grad strides do not match bucket view strides. "
+                        "This may indicate grad was not created according to the "
+                        "gradient layout contract, or that the param's strides "
+                        "changed since DDP was constructed.  This is not an error, "
+                        "but may impair performance.\n"
+                        "grad.sizes() = ", grad.sizes(),
+                        ", strides() = ", grad.strides(), "\n",
+                        "bucket_view.sizes() = ", bucket_view.sizes(),
+                        "bucket_view.strides() = ", bucket_view.strides());
+      }
+      // imitates wrapped_scalar_tensor in ATen/native/BinaryOps.cpp
+      auto wrapped = c10::scalar_to_tensor(double(1.)/process_group_->getSize());
+      wrapped.unsafeGetTensorImpl()->set_wrapped_number(true);
+      // Divides while copying into the bucket view.
+      at::native::mul_out(bucket_view, grad, wrapped);
```

Comment:
@mrshenli are the diffs here ok?  grad is now divided while being copied into bucket_view, but it's still not being modified, so I think it's ok.
A more subtle thing is that with this PR's other diffs, grad may no longer be rowmajor contiguous.  I'm not sure how that affects `dist_autograd` in general if the callback ends up routing into the `dist_autograd` context.

Reply:
This LGTM. 

Curious, what difference does `set_wrapped_number` make? Is it like `mul_out` only support scalar tensor or there is a shortcut for this?

Regarding dist autograd, as DDP + RPC is a very new feature and will be released as beta. It still need more tests and apps to verify and try it out. I think even if it does not work with this PR, it should not block us from landing this. We can fix that in follow PRs.

Reply:
`at::native::mul_out` only has an overload that accepts three Tensor args, so I need to convert the scalar to a tensor.

I'm not sure exactly why `set_wrapped_number` in particular is needed.  I don't think it's needed for TensorIterator kernels inside mul_out, because the [lambda capturing](https://github.com/pytorch/pytorch/blob/0b90b9cdd36de3131e7b3868c71acc48170f3f1a/aten/src/ATen/native/cuda/Loops.cuh#L88-L104) of CPU scalars for GPU kernels is based on [TensorIterator::is_cpu_scalar](https://github.com/pytorch/pytorch/blob/9ca7fdcef083f78692cf8ecc68e32472b677a905/aten/src/ATen/native/TensorIterator.cpp#L607-L619) which does not rely on Tensor::is_wrapped_number.  I included `set_wrapped_number` because it seems to be the standard practice when converting scalars to Tensors (e.g. in [BinaryOps.cpp](https://github.com/pytorch/pytorch/blob/9ca7fdcef083f78692cf8ecc68e32472b677a905/aten/src/ATen/native/BinaryOps.cpp#L279-L283) and [python_arg_parser.cpp](https://github.com/pytorch/pytorch/blob/9ca7fdcef083f78692cf8ecc68e32472b677a905/torch/csrc/utils/python_arg_parser.cpp#L844-L846)).